Sources/NIOTCPEchoClient/Client.swift

```diff
@@ -0,0 +1,91 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if swift(>=5.9)
+@_spi(AsyncChannel) import NIOCore
+@_spi(AsyncChannel) import NIOPosix
+
+@available(macOS 14, *)
+@main
+struct Client {
+    /// The host to connect to.
+    private let host: String
+    /// The port to connect to.
+    private let port: Int
+    /// The client's event loop group.
+    private let eventLoopGroup: any EventLoopGroup
+
+    static func main() async throws {
+        // We are creating this event loop group at the top level and it will live until
+        // the process exits. This means we also don't have to shut it down.
+        //
+        // Note that we start this group with 1 thread. In general, most NIO programs
+        // should use 1 thread as a default unless they're planning to be a network
+        // proxy or something that is expecting to be dominated by packet parsing. Most
+        // servers aren't, and NIO is very fast, so 1 NIO thread is quite capable of
+        // saturating the average small to medium sized machine.
+        let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```

Comment:
Sure, in this case you don't have to shut it down but why leak stuff if it's trivial to not leak?

`defer { try! elg.syncShutdownGracefully() }`

Reply:
So moving this into a global variable and naming it `globalSingletonGroup` or so would be much better because copy & paste wouldn't lead to super bad leaks. The code as is is just incorrect because `eventLoopGroup` will get `deinit`ed and it's illegal to `deinit` this without shutting it down (don't think that's enforced today though).

Thinking about it more: I've long come to the conclusion that NIO needs to provides singletons for users who don't need tight control and I think https://github.com/apple/swift-nio/pull/2471 is the answer :)

Reply:
Holding this PR until we land the singleton work. Thanks @weissi!